python/pip_install/extract_wheels/lib/requirements.py

```diff
@@ -23,6 +23,14 @@ def parse_extras(requirements_path: str) -> Dict[str, Set[str]]:
     return extras_requested
 
 
+def sanitise_requirement(requirement: str) -> str:
+    """Given a requirement string, returns a normalized version of the requirement name.
+    The case and use of "_" vs "-" in a requirements file may vary.
+    https://peps.python.org/pep-0426/#name
+    """
+    return requirement.replace("-", "_").lower()
+
+
```

Comment:
Thanks for digging that up @jvolkman.
Taking the regex from PEP 503 seems like a good fit.

I should probably update the name of the PR to reflect the more accurate PEP.

Reply:
Yes, PEP-426 is `Withdrawn`. [PEP-508](https://peps.python.org/pep-0508/#names) or [](https://peps.python.org/pep-0503/#normalized-names) is what we should be referencing here.

We are in full control of the code. So normalizing when creating the mapping between `binary distributions (wheels)-> extras requested` should be done when the requirementst.txt is parsed, and when installing such a wheel and setting up the relevant dependencies for the extras.

Reply:
I checked through `pypa/installer` and didn't find a method for normalizing the distribution name while parsing wheel metadata:
- https://github.com/pypa/installer/issues/97

However, I did find that `pypa/pkg_resources` provides an interface for normalization:
- https://setuptools.pypa.io/en/latest/pkg_resources.html#requirement-objects

I applied this to the `.name` property on `Wheel`.

I had to apply the normalization in `_parse_requirement_for_extra` because that's the closest spot where the distribution name is isolated from the `extras`.

`pkg_resources.Requirement` can also parse the `extras` section but that applies `safe_extra` to each entry.
That _may_ be a suitable replacement for the current regex in `_parse_requirement_for_extra`, as long as "fixing" the extras doesn't break their inclusion.

Reply:
Took a look at what piptools was doing. They use the censored versioned of `packaging` to normalize the name https://github.com/jazzband/pip-tools/blob/9fb2d425d6d7abb9d9b9c51ebb711af420da25f4/piptools/utils.py#L30